Reregister event hub consumer after unregister caused IllegalStateException

# Steps to reproduce:
1. when the microservice starts all binders go up and … all is ok
2. after several tasks I decide to stop the binders and … all is ok
3. after some while I decide to restart the binders….I have a problem

# Stacktrace as below:
```
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.lang.IllegalStateException: Register cannot be called on an EventProcessorHost after unregister. Please create a new EventProcessorHost instance.
Caused by: java.lang.IllegalStateException: Register cannot be called on an EventProcessorHost after unregister. Please create a new EventProcessorHost instance.
at com.microsoft.azure.eventprocessorhost.EventProcessorHost.registerEventProcessorFactory(EventProcessorHost.java:461)
at com.microsoft.azure.spring.integration.eventhub.impl.AbstractEventHubTemplate.register(AbstractEventHubTemplate.java:103)
at com.microsoft.azure.spring.integration.eventhub.impl.EventHubTemplate.subscribe(EventHubTemplate.java:40)
at com.microsoft.azure.spring.integration.core.api.SubscribeByGroupOperation.subscribe(SubscribeByGroupOperation.java:30)
at com.microsoft.azure.spring.integration.core.AbstractInboundChannelAdapter.doStart(AbstractInboundChannelAdapter.java:38)
at org.springframework.integration.endpoint.AbstractEndpoint.start(AbstractEndpoint.java:156)
at org.springframework.cloud.stream.binder.DefaultBinding.start(DefaultBinding.java:116)
at com.wba.oneit.oneleo.config.BinderStartStopService.lambda$start$0(BinderStartStopService.java:41)
at java.util.concurrent.ConcurrentHashMap$EntrySetView.forEach(ConcurrentHashMap.java:4795)
at com.wba.oneit.oneleo.config.BinderStartStopService.start(BinderStartStopService.java:39)
```
# The root cause is inside:
```
protected void register(String name, String consumerGroup, EventHubProcessor eventProcessor) {
        EventProcessorHost host = this.clientFactory.getOrCreateEventProcessorHost(name, consumerGroup);
        host.registerEventProcessorFactory(context -> eventProcessor, buildEventProcessorOptions(startPosition));
    }

    protected void unregister(String name, String consumerGroup) {
        this.clientFactory.getOrCreateEventProcessorHost(name, consumerGroup).unregisterEventProcessor()
                          .whenComplete((s, t) -> {
                              if (t != null) {
                                  log.warn(String.format("Failed to unregister consumer '%s' with group '%s'", name,
                                          consumerGroup), t);
                              }
                          });
```
where this.clientFactory.getOrCreateEventProcessorHost is called in both register and unregister cases. 

But when I do the mentioned sequence of operations (register , unregister,  register)  I have an error during the second register phase because the EventProcessorHost retrieved from the map 
is registered and it has valorised variables that are null during the first register phase .

# How to fix

It seems that we should remove `EventProcessorHost` after unregister, then creating new one after reregister.

--- a/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/api/EventHubClientFactory.java
+++ b/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/api/EventHubClientFactory.java
@@ -6,12 +6,15 @@
 
 package com.microsoft.azure.spring.integration.eventhub.api;
 
+import java.util.Optional;
+
 import com.microsoft.azure.eventhubs.EventHubClient;
 import com.microsoft.azure.eventhubs.PartitionSender;
 import com.microsoft.azure.eventprocessorhost.EventProcessorHost;
 
 /**
  * @author Warren Zhu
+ * @author Domenico Sibilio
  */
 public interface EventHubClientFactory {
 
@@ -20,5 +23,9 @@ public interface EventHubClientFactory {
     PartitionSender getOrCreatePartitionSender(String eventhub, String partition);
 
     EventProcessorHost getOrCreateEventProcessorHost(String name, String consumerGroup);
+
+    Optional<EventProcessorHost> getEventProcessorHost(String name, String consumerGroup);
+
+    EventProcessorHost removeEventProcessorHost(String name, String consumerGroup);
 
 }

--- a/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/factory/DefaultEventHubClientFactory.java
+++ b/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/factory/DefaultEventHubClientFactory.java
@@ -6,6 +6,15 @@
 
 package com.microsoft.azure.spring.integration.eventhub.factory;
 
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
 import com.microsoft.azure.eventhubs.EventHubClient;
 import com.microsoft.azure.eventhubs.EventHubException;
 import com.microsoft.azure.eventhubs.PartitionSender;
@@ -18,17 +27,10 @@ import com.microsoft.azure.spring.integration.eventhub.impl.EventHubRuntimeExcep
 import com.microsoft.azure.spring.integration.eventhub.util.HostnameHelper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
 import org.springframework.beans.factory.DisposableBean;
 import org.springframework.lang.NonNull;
 import org.springframework.util.Assert;
-
-import java.io.IOException;
-import java.util.Map;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.Executors;
-import java.util.function.BiFunction;
-import java.util.function.Function;
 
 /**
  * Default implementation of {@link EventHubClientFactory}.
@@ -90,10 +92,10 @@ public class DefaultEventHubClientFactory implements EventHubClientFactory, Disp
 
     private <K, V> void close(Map<K, V> map, Function<V, CompletableFuture<Void>> close) {
         CompletableFuture.allOf(map.values().stream().map(close).toArray(CompletableFuture[]::new))
-                         .exceptionally((ex) -> {
-                             log.warn("Failed to clean event hub client factory", ex);
-                             return null;
-                         });
+        .exceptionally((ex) -> {
+            log.warn("Failed to clean event hub client factory", ex);
+            return null;
+        });
     }
 
     @Override
@@ -116,5 +118,15 @@ public class DefaultEventHubClientFactory implements EventHubClientFactory, Disp
     @Override
     public EventProcessorHost getOrCreateEventProcessorHost(String name, String consumerGroup) {
         return this.processorHostCreator.apply(name, consumerGroup);
+    }
+
+    @Override
+    public Optional<EventProcessorHost> getEventProcessorHost(String name, String consumerGroup) {
+        return Optional.ofNullable(this.processorHostMap.get(Tuple.of(name, consumerGroup)));
+    }
+
+    @Override
+    public EventProcessorHost removeEventProcessorHost(String name, String consumerGroup) {
+        return this.processorHostMap.remove(Tuple.of(name, consumerGroup));
     }
 }

--- a/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/impl/AbstractEventHubTemplate.java
+++ b/spring-integration-azure/spring-integration-eventhubs/src/main/java/com/microsoft/azure/spring/integration/eventhub/impl/AbstractEventHubTemplate.java
@@ -6,6 +6,20 @@
 
 package com.microsoft.azure.spring.integration.eventhub.impl;
 
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.NonNull;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
 import com.google.common.base.Strings;
 import com.microsoft.azure.eventhubs.EventData;
 import com.microsoft.azure.eventhubs.EventHubClient;
@@ -18,15 +32,6 @@ import com.microsoft.azure.spring.integration.core.api.PartitionSupplier;
 import com.microsoft.azure.spring.integration.core.api.StartPosition;
 import com.microsoft.azure.spring.integration.eventhub.api.EventHubClientFactory;
 import com.microsoft.azure.spring.integration.eventhub.converter.EventHubMessageConverter;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-import org.springframework.lang.NonNull;
-import org.springframework.messaging.Message;
-import org.springframework.util.Assert;
-
-import java.util.*;
-import java.util.concurrent.CompletableFuture;
-import java.util.stream.Collectors;
 
 /**
  * Abstract base implementation of event hub template.
@@ -104,13 +109,22 @@ public class AbstractEventHubTemplate {
     }
 
     protected void unregister(String name, String consumerGroup) {
-        this.clientFactory.getOrCreateEventProcessorHost(name, consumerGroup).unregisterEventProcessor()
-                          .whenComplete((s, t) -> {
-                              if (t != null) {
-                                  log.warn(String.format("Failed to unregister consumer '%s' with group '%s'", name,
-                                          consumerGroup), t);
-                              }
-                          });
+        this.clientFactory
+        .getEventProcessorHost(name, consumerGroup)
+        .ifPresent(eventProcessorHost -> unregisterEventProcessor(eventProcessorHost, name, consumerGroup));
+    }
+
+    private void unregisterEventProcessor(EventProcessorHost eventProcessorHost, String name, String consumerGroup) {
+        this.clientFactory.removeEventProcessorHost(name, consumerGroup);
+
+        eventProcessorHost
+        .unregisterEventProcessor()
+        .whenComplete((s, t) -> {
+            if (t != null) {
+                log.warn(String.format("Failed to unregister consumer '%s' with group '%s'", name,
+                        consumerGroup), t);
+            }
+        });
     }
 
     protected Map<String, Object> buildPropertiesMap() {

--- a/spring-integration-azure/spring-integration-eventhubs/src/test/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactoryTest.java
+++ b/spring-integration-azure/spring-integration-eventhubs/src/test/java/com/microsoft/azure/spring/integration/eventhub/DefaultEventHubClientFactoryTest.java
@@ -6,6 +6,8 @@
 
 package com.microsoft.azure.spring.integration.eventhub;
 
+import java.util.Optional;
+
 import com.microsoft.azure.AzureEnvironment;
 import com.microsoft.azure.eventhubs.EventHubClient;
 import com.microsoft.azure.eventhubs.PartitionSender;
@@ -18,7 +20,6 @@ import com.microsoft.azure.spring.integration.eventhub.api.EventHubClientFactory
 import com.microsoft.azure.spring.integration.eventhub.factory.DefaultEventHubClientFactory;
 import com.microsoft.azure.spring.integration.eventhub.factory.EventHubConnectionStringProvider;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.mockito.Mock;
@@ -26,13 +27,16 @@ import org.powermock.api.mockito.PowerMockito;
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.*;
 import static org.mockito.ArgumentMatchers.*;
-import static org.powermock.api.mockito.PowerMockito.when;
+import static org.mockito.Mockito.*;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+import static org.powermock.api.mockito.PowerMockito.*;
 
 @RunWith(PowerMockRunner.class)
-@PrepareForTest({EventHubClient.class, StorageConnectionStringProvider.class, EventProcessorHost.class})
+@PrepareForTest({EventHubClient.class, StorageConnectionStringProvider.class,
+    EventProcessorHost.class, DefaultEventHubClientFactory.class })
 public class DefaultEventHubClientFactoryTest {
 
     @Mock
@@ -55,12 +59,11 @@ public class DefaultEventHubClientFactoryTest {
 
     @Mock
     EventHubConnectionStringProvider connectionStringProvider;
-    
+
     @Mock
     StorageConnectionStringProvider storageConnectionStringProvider;
 
     private EventHubClientFactory clientFactory;
-    private String checkpointStorageAccount = "sa";
     private String eventHubName = "eventHub";
     private String consumerGroup = "group";
     private String connectionString = "conStr";
@@ -76,15 +79,15 @@ public class DefaultEventHubClientFactoryTest {
         PowerMockito.mockStatic(StorageConnectionStringProvider.class);
         when(StorageConnectionStringProvider
                 .getConnectionString(isA(StorageAccount.class), isA(AzureEnvironment.class)))
-                .thenReturn(connectionString);
+        .thenReturn(connectionString);
         when(StorageConnectionStringProvider
                 .getConnectionString(isA(StorageAccount.class), isA(AzureEnvironment.class)))
-                .thenReturn(connectionString);
+        .thenReturn(connectionString);
         when(resourceManagerProvider.getStorageAccountManager()).thenReturn(storageAccountManager);
         when(storageAccountManager.getOrCreate(any())).thenReturn(storageAccount);
         PowerMockito.whenNew(EventProcessorHost.class).withAnyArguments().thenReturn(eventProcessorHost);
 
-        this.clientFactory = new DefaultEventHubClientFactory(connectionStringProvider, connectionString);
+        this.clientFactory = spy(new DefaultEventHubClientFactory(connectionStringProvider, connectionString));
     }
 
     @Test
@@ -104,11 +107,49 @@ public class DefaultEventHubClientFactoryTest {
     }
 
     @Test
-    @Ignore("Cannot mock EventProcessorHost constructor")
-    public void testGetEventProcessorHost() {
+    public void testGetEventProcessorHost() throws Exception {
+        clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
+        Optional<EventProcessorHost> optionalEph = clientFactory.getEventProcessorHost(eventHubName, consumerGroup);
+
+        assertTrue(optionalEph.isPresent());
+    }
+
+    @Test
+    public void testGetNullEventProcessorHost() {
+        Optional<EventProcessorHost> optionalEph = clientFactory.getEventProcessorHost(eventHubName, consumerGroup);
+        assertFalse(optionalEph.isPresent());
+    }
+
+    @Test
+    public void testRemoveEventProcessorHost() {
+        EventProcessorHost host = clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
+        EventProcessorHost another = clientFactory.removeEventProcessorHost(eventHubName, consumerGroup);
+
+        assertSame(host, another);
+    }
+
+    @Test
+    public void testRemoveAbsentEventProcessorHost() {
+        EventProcessorHost eventProcessorHost = clientFactory.removeEventProcessorHost(eventHubName, consumerGroup);
+        assertNull(eventProcessorHost);
+    }
+
+    @Test
+    public void testGetOrCreateEventProcessorHost() throws Exception {
         EventProcessorHost host = clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
         assertNotNull(host);
-        EventProcessorHost another = clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
-        assertEquals(host, another);
+        clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
+        
+        verifyPrivate(clientFactory).invoke("createEventProcessorHost", eventHubName, consumerGroup);
+    }
+    
+    @Test
+    public void testRecreateEventProcessorHost() throws Exception {
+        EventProcessorHost host = clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
+        assertNotNull(host);
+        clientFactory.removeEventProcessorHost(eventHubName, consumerGroup);
+        clientFactory.getOrCreateEventProcessorHost(eventHubName, consumerGroup);
+        
+        verifyPrivate(clientFactory, times(2)).invoke("createEventProcessorHost", eventHubName, consumerGroup);
     }
 }

--- a/spring-integration-azure/spring-integration-eventhubs/src/test/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplateSubscribeTest.java
+++ b/spring-integration-azure/spring-integration-eventhubs/src/test/java/com/microsoft/azure/spring/integration/eventhub/EventHubTemplateSubscribeTest.java
@@ -6,6 +6,9 @@
 
 package com.microsoft.azure.spring.integration.eventhub;
 
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
 import com.microsoft.azure.eventprocessorhost.EventProcessorHost;
 import com.microsoft.azure.eventprocessorhost.EventProcessorOptions;
 import com.microsoft.azure.eventprocessorhost.IEventProcessorFactory;
@@ -18,9 +21,7 @@ import org.junit.runner.RunWith;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
 
-import java.util.concurrent.CompletableFuture;
-
-import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.ArgumentMatchers.*;
 import static org.mockito.Mockito.*;
 
 @RunWith(MockitoJUnitRunner.class)
@@ -38,6 +39,7 @@ public class EventHubTemplateSubscribeTest extends SubscribeByGroupOperationTest
         future.complete(null);
         this.subscribeByGroupOperation = new EventHubTemplate(mockClientFactory);
         when(this.mockClientFactory.getOrCreateEventProcessorHost(anyString(), anyString())).thenReturn(this.host);
+        when(this.mockClientFactory.getEventProcessorHost(anyString(), anyString())).thenReturn(Optional.of(this.host));
         when(this.host
                 .registerEventProcessorFactory(isA(IEventProcessorFactory.class), isA(EventProcessorOptions.class)))
                 .thenReturn(future);
@@ -62,6 +64,7 @@ public class EventHubTemplateSubscribeTest extends SubscribeByGroupOperationTest
 
     @Override
     protected void verifySubscriberUnregistered(int times) {
+        verify(this.mockClientFactory, times(times)).removeEventProcessorHost(anyString(), anyString());
         verify(this.host, times(times)).unregisterEventProcessor();
     }
 }

